### DISCUSSION TITLE
Hi, how to simulate the effect of flowing water along a specified pipe?

### DISCUSSION THREAD

**Discussion by shenshengyi** [iTwin/itwinjs-core#1008] [Q&A] [answered]
Hi, I want to simulate the effect of flowing water along the specified pipeline. Can you give me a suggestion? Thanks.

**Comment by pmconne** [comment, marked as answer]
You are going to need to be more specific.

**Reply by shenshengyi** [reply]
> 
> 
> You are going to need to be more specific.

hi, there are many long water pipes in the model. I hope to select a water pipe and simulate a water flow along the water pipe. I now want to use decorators and animations to simulate it. Do you have any good suggestions? Thank you.

**Comment by pmconne** [comment]
Do you know what you want your decorations to look like and are asking how to implement it? If so, tell us what you want it to look like. Or, are you asking for suggestions for what it should look like? If so, tell us what you want to simulate - direction? pressure? etc?

I presume you will need to obtain geometry from your pipe elements. Are you able to do that or do you need a way to do that? How is the pipe geometry represented - curve vectors? BReps? cylinders?

**Reply by shenshengyi** [reply]
hi, this is an example I made with decoration and animation, solid may be better.
After I got the geometry of the pipe, I didn’t know how to proceed. I could only reconstruct a decorative geometry based on the geometric parameters of the pipe, and then approximate the flow of water along the pipe. Thanks.

https://user-images.githubusercontent.com/26344907/112074136-895cd700-8bb0-11eb-9159-75c6ac25705a.mp4

**Reply by pmconne** [reply]
I see 15 seconds of blackness. I am investing effort into answering your questions. Would you mind investing some in asking them please? Where is your code? What do you want to see?

**Reply by shenshengyi** [reply]
> 
> 
> I see 15 seconds of blackness. I am investing effort into answering your questions. Would you mind investing some in asking them please? Where is your code? What do you want to see?

hi, in the current scene, there are many very long pipelines. I hope to select one of the specified pipelines, and then simulate the trajectory of this pipeline in the form of water flow.

Now I use decorators and animations to make a simple effect using some geometric location points of the pipeline, and I am not satisfied.

**I would like to ask, is it possible to create a texture on the front end of itwinsjs, and then use animation to move the texture coordinates? The purpose is to visually simulate the effect of water flow along the pipeline.**  Thanks

**Reply by pmconne** [reply]
[RenderSystem](https://github.com/imodeljs/imodeljs/blob/ad5e3d840707a464d151cdf8cec0cfc8b428fd22/core/frontend/src/render/RenderSystem.ts#L434) has various methods for creating textures on the frontend.
Once you have created a RenderGraphic you cannot change things like its texture coordinates. You would need to produce a new RenderGraphic with different texture coordinates.

**Reply by shenshengyi** [reply]
Thank you very much, your suggestions helped me a lot.
I will use this method to fulfill my needs.
